{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem\n",
    "\n",
    "Either strand of a DNA double helix can serve as the coding strand for RNA transcription. Hence, a given DNA string implies six total reading frames, or ways in which the same region of DNA can be translated into amino acids: three reading frames result from reading the string itself, whereas three more result from reading its reverse complement.\n",
    "\n",
    "An open reading frame (ORF) is one which starts from the start codon and ends by stop codon, without any other stop codons in between. Thus, a candidate protein string is derived by translating an open reading frame into amino acids until a stop codon is reached.\n",
    "\n",
    "Given: A DNA string s\n",
    "\n",
    "of length at most 1 kbp in FASTA format.\n",
    "\n",
    "Return: Every distinct candidate protein string that can be translated from ORFs of s\n",
    "\n",
    ". Strings can be returned in any order.\n",
    "Sample Dataset\n",
    "\n",
    ">Rosalind_99\n",
    "AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG\n",
    "\n",
    "Sample Output\n",
    "\n",
    "MLLGSFRLIPKETLIQVAGSSPCNLS\n",
    "M\n",
    "MGMTPRLGLESLLE\n",
    "MTPRLGLESLLE\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ATGTAA', 'ATGTAA']\n"
     ]
    }
   ],
   "source": [
    "DNA = 'AAAATGATGTAAATGTAAAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG'\n",
    "x = 0\n",
    "y = 3\n",
    "ORFs = []\n",
    "while y < len(DNA):\n",
    "    ORF_DNA = ''\n",
    "    if DNA[x:y] == 'ATG':\n",
    "        ORF_DNA = ORF_DNA + DNA[x:y]\n",
    "        x = x + 3\n",
    "        y = y + 3\n",
    "        j = x\n",
    "    if DNA[x:y] != 'ATG':\n",
    "        if ORF_DNA != '':\n",
    "            ORF_DNA = ORF_DNA + DNA[x:y]\n",
    "            if DNA[x:y] in ['TAG','TAA','TGA']:\n",
    "                ORFs.append(ORF_DNA)\n",
    "                x = j\n",
    "                y = x + 3\n",
    "                ORF_DNA = ''\n",
    "                continue\n",
    "        else:\n",
    "            x = x + 3\n",
    "            y = y + 3\n",
    "            continue\n",
    "                \n",
    "            \n",
    "print (ORFs)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DNA = 'AAAATGATGTAAATGTAAAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG'\n",
    "x = 0\n",
    "y = 3\n",
    "ORFs = []\n",
    "while y < len(DNA):\n",
    "    ORF_DNA = ''\n",
    "    if DNA[x:y] == 'ATG':\n",
    "        ORF_DNA = ORF_DNA + DNA[x:y]\n",
    "        x = x + 3\n",
    "        y = y + 3\n",
    "        j = x\n",
    "    if DNA[x:y] != 'ATG':\n",
    "        if DNA[x:y] in ['TAG','TAA','TGA']:\n",
    "            ORF_DNA = ORF_DNA + DNA[x:y]\n",
    "            ORFs.append(ORF_DNA)\n",
    "            x = j\n",
    "            y = x + 3\n",
    "            ORF_DNA = ''\n",
    "            continue\n",
    "        elif ORF_DNA != '':\n",
    "            ORF_DNA = ORF_DNA + DNA[x:y]\n",
    "            x = x + 3\n",
    "            y = y + 3\n",
    "            continue\n",
    "        else:\n",
    "            x = x + 3\n",
    "            y = y + 3\n",
    "            continue\n",
    "                \n",
    "            \n",
    "print (ORFs)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AAT', 'GAT', 'GTA', 'AAT', 'GTA', 'AAA', 'CTC', 'AGG', 'TTA', 'CAT', 'GGG', 'GAT', 'GAC', 'CCC', 'GCG', 'ACT', 'TGG', 'ATT', 'AGA', 'GTC', 'TCT', 'TTT', 'GGA', 'ATA', 'AGC', 'CTG', 'AAT', 'GAT', 'CCG', 'AGT', 'AGC', 'ATC']\n",
      "{1: ['AAA', 'ATG', 'ATG', 'TAA', 'ATG', 'TAA', 'AAC', 'TCA', 'GGT', 'TAC', 'ATG', 'GGG', 'ATG', 'ACC', 'CCG', 'CGA', 'CTT', 'GGA', 'TTA', 'GAG', 'TCT', 'CTT', 'TTG', 'GAA', 'TAA', 'GCC', 'TGA', 'ATG', 'ATC', 'CGA', 'GTA', 'GCA', 'TCT'], 2: ['AAA', 'TGA', 'TGT', 'AAA', 'TGT', 'AAA', 'ACT', 'CAG', 'GTT', 'ACA', 'TGG', 'GGA', 'TGA', 'CCC', 'CGC', 'GAC', 'TTG', 'GAT', 'TAG', 'AGT', 'CTC', 'TTT', 'TGG', 'AAT', 'AAG', 'CCT', 'GAA', 'TGA', 'TCC', 'GAG', 'TAG', 'CAT'], 3: ['AAT', 'GAT', 'GTA', 'AAT', 'GTA', 'AAA', 'CTC', 'AGG', 'TTA', 'CAT', 'GGG', 'GAT', 'GAC', 'CCC', 'GCG', 'ACT', 'TGG', 'ATT', 'AGA', 'GTC', 'TCT', 'TTT', 'GGA', 'ATA', 'AGC', 'CTG', 'AAT', 'GAT', 'CCG', 'AGT', 'AGC', 'ATC']}\n"
     ]
    }
   ],
   "source": [
    "DNA = 'AAAATGATGTAAATGTAAAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG'\n",
    "\n",
    "\"\"\"\n",
    "code to get a DNA sequence into a dictionary of reading frames containing codons \n",
    "these can then be iterated through\n",
    "\"\"\"\n",
    "x = 0\n",
    "y = 3\n",
    "j = 1\n",
    "ORFs = {}\n",
    "codons = []\n",
    "while y <= (len(DNA)):\n",
    "    codons.append(DNA[x:y])\n",
    "    x = x + 3\n",
    "    y = y + 3\n",
    "    if y >= len(DNA)-2:\n",
    "        ORFs[j] = codons\n",
    "        if j < 3:\n",
    "            codons = []\n",
    "            x = 0 + j\n",
    "            y = 3 + j\n",
    "            j = j + 1\n",
    "            continue\n",
    "        else:\n",
    "            break\n",
    "print (codons)\n",
    "print (ORFs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTGAGATGCATCATTTT\n"
     ]
    }
   ],
   "source": [
    "sequence = 'AAAATGATGCATCTCAG'\n",
    "def reverse_complement(DNA):\n",
    "    revcompDict = {'A':'T', 'C':'G', 'T':'A', 'G':'C'}\n",
    "    i = 1\n",
    "    x = 0\n",
    "    revseq = ''\n",
    "    revcompseq = ''\n",
    "    for i in range (1, len(DNA) + 1):\n",
    "        revseq = DNA[-i]\n",
    "        revcompseq = revcompseq + revcompDict[revseq]\n",
    "    print (revcompseq)\n",
    "\n",
    "reverse_complement(sequence)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: ['K',\n",
       "  'M',\n",
       "  'M',\n",
       "  '_',\n",
       "  'M',\n",
       "  '_',\n",
       "  'N',\n",
       "  'S',\n",
       "  'G',\n",
       "  'Y',\n",
       "  'M',\n",
       "  'G',\n",
       "  'M',\n",
       "  'T',\n",
       "  'P',\n",
       "  'R',\n",
       "  'L',\n",
       "  'G',\n",
       "  'L',\n",
       "  'E',\n",
       "  'S',\n",
       "  'L',\n",
       "  'L',\n",
       "  'E',\n",
       "  '_',\n",
       "  'A',\n",
       "  '_',\n",
       "  'M',\n",
       "  'I',\n",
       "  'R',\n",
       "  'V',\n",
       "  'A',\n",
       "  'S'],\n",
       " 2: ['K',\n",
       "  '_',\n",
       "  'C',\n",
       "  'K',\n",
       "  'C',\n",
       "  'K',\n",
       "  'T',\n",
       "  'Q',\n",
       "  'V',\n",
       "  'T',\n",
       "  'W',\n",
       "  'G',\n",
       "  '_',\n",
       "  'P',\n",
       "  'R',\n",
       "  'D',\n",
       "  'L',\n",
       "  'D',\n",
       "  '_',\n",
       "  'S',\n",
       "  'L',\n",
       "  'F',\n",
       "  'W',\n",
       "  'N',\n",
       "  'K',\n",
       "  'P',\n",
       "  'E',\n",
       "  '_',\n",
       "  'S',\n",
       "  'E',\n",
       "  '_',\n",
       "  'H'],\n",
       " 3: ['N',\n",
       "  'D',\n",
       "  'V',\n",
       "  'N',\n",
       "  'V',\n",
       "  'K',\n",
       "  'L',\n",
       "  'R',\n",
       "  'L',\n",
       "  'H',\n",
       "  'G',\n",
       "  'D',\n",
       "  'D',\n",
       "  'P',\n",
       "  'A',\n",
       "  'T',\n",
       "  'W',\n",
       "  'I',\n",
       "  'R',\n",
       "  'V',\n",
       "  'S',\n",
       "  'F',\n",
       "  'G',\n",
       "  'I',\n",
       "  'S',\n",
       "  'L',\n",
       "  'N',\n",
       "  'D',\n",
       "  'P',\n",
       "  'S',\n",
       "  'S',\n",
       "  'I'],\n",
       " 4: ['L',\n",
       "  'R',\n",
       "  'C',\n",
       "  'Y',\n",
       "  'S',\n",
       "  'D',\n",
       "  'H',\n",
       "  'S',\n",
       "  'G',\n",
       "  'L',\n",
       "  'F',\n",
       "  'Q',\n",
       "  'K',\n",
       "  'R',\n",
       "  'L',\n",
       "  '_',\n",
       "  'S',\n",
       "  'K',\n",
       "  'S',\n",
       "  'R',\n",
       "  'G',\n",
       "  'H',\n",
       "  'P',\n",
       "  'H',\n",
       "  'V',\n",
       "  'T',\n",
       "  '_',\n",
       "  'V',\n",
       "  'L',\n",
       "  'H',\n",
       "  'L',\n",
       "  'H',\n",
       "  'H'],\n",
       " 5: ['D',\n",
       "  'A',\n",
       "  'T',\n",
       "  'R',\n",
       "  'I',\n",
       "  'I',\n",
       "  'Q',\n",
       "  'A',\n",
       "  'Y',\n",
       "  'S',\n",
       "  'K',\n",
       "  'R',\n",
       "  'D',\n",
       "  'S',\n",
       "  'N',\n",
       "  'P',\n",
       "  'S',\n",
       "  'R',\n",
       "  'G',\n",
       "  'V',\n",
       "  'I',\n",
       "  'P',\n",
       "  'M',\n",
       "  '_',\n",
       "  'P',\n",
       "  'E',\n",
       "  'F',\n",
       "  'Y',\n",
       "  'I',\n",
       "  'Y',\n",
       "  'I'],\n",
       " 6: ['M',\n",
       "  'L',\n",
       "  'L',\n",
       "  'G',\n",
       "  'S',\n",
       "  'F',\n",
       "  'R',\n",
       "  'L',\n",
       "  'I',\n",
       "  'P',\n",
       "  'K',\n",
       "  'E',\n",
       "  'T',\n",
       "  'L',\n",
       "  'I',\n",
       "  'Q',\n",
       "  'V',\n",
       "  'A',\n",
       "  'G',\n",
       "  'S',\n",
       "  'S',\n",
       "  'P',\n",
       "  'C',\n",
       "  'N',\n",
       "  'L',\n",
       "  'S',\n",
       "  'F',\n",
       "  'T',\n",
       "  'F',\n",
       "  'T',\n",
       "  'S']}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence = 'AAAATGATGTAAATGTAAAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG'\n",
    "\n",
    "\"\"\"\n",
    "code to get a DNA sequence into a dictionary of reading frames containing codons \n",
    "these can then be iterated through attempting to create with no dependencies such as biopython.\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "reverse complement function- provides the reverse_complement when provided a DNA input as an argument\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def reverse_complement(DNA):\n",
    "    DNA = DNA.upper()\n",
    "    revcompDict = {'A':'T', 'C':'G', 'T':'A', 'G':'C'}\n",
    "    i = 1\n",
    "    x = 0\n",
    "    revseq = ''\n",
    "    revcompseq = ''\n",
    "    for i in range (1, len(DNA) + 1):\n",
    "        revseq = DNA[-i]\n",
    "        revcompseq = revcompseq + revcompDict[revseq]\n",
    "    return revcompseq\n",
    "\n",
    "reverse_complement(sequence)\n",
    "\n",
    "\"\"\"\n",
    "CODON_dictionary function- takes a DNA sequence as an input and provides a list of codons in ALL reading frames of a sequence including\n",
    "reverse complement- outputs to a dictionary of form ORF[i]=['AAA','ATG', etc] where i is the reading frame\n",
    "the reverse codons start from 4 and go to 6\n",
    "\"\"\"\n",
    "\n",
    "def CODON_LIST(DNA):\n",
    "    DNA = DNA.upper()\n",
    "    x = 0\n",
    "    y = 3\n",
    "    j = 1\n",
    "    ORFs = {}\n",
    "    codons = []\n",
    "    while y <= (len(DNA)):\n",
    "        codons.append(DNA[x:y])\n",
    "        x = x + 3\n",
    "        y = y + 3\n",
    "        if y >= len(DNA)-2:\n",
    "            ORFs[j] = codons\n",
    "            if j < 3:\n",
    "                codons = []\n",
    "                x = 0 + j\n",
    "                y = 3 + j\n",
    "                j = j + 1\n",
    "                continue\n",
    "            else:\n",
    "                break\n",
    "    \n",
    "    DNA = reverse_complement(DNA)\n",
    "    x = 0\n",
    "    y = 3\n",
    "    j = 4\n",
    "    codons = []\n",
    "    while y <= (len(DNA)):\n",
    "        codons.append(DNA[x:y])\n",
    "        x = x + 3\n",
    "        y = y + 3\n",
    "        if y >= len(DNA)-2:\n",
    "            ORFs[j] = codons\n",
    "            if j < 6:\n",
    "                codons = []\n",
    "                x = 0 + j\n",
    "                y = 3 + j\n",
    "                j = j + 1\n",
    "                continue\n",
    "            else:\n",
    "                break\n",
    "    return (ORFs)\n",
    "\n",
    "\"\"\"\n",
    "the translate function is designed to work on any LIST OF CODONS in a dictionary format where the key is an \n",
    "integer representing the reading frame from 1 - 6. therefore it has the function to check if the reading frame\n",
    "exists in the dictionary, translate the sequences to protein and output a protein sequence dictionary.\n",
    "\"\"\"\n",
    "CODON_LIST = CODON_LIST(sequence)\n",
    "\n",
    "def TRANSLATE(CODON_LIST):\n",
    "    protein_frame_dict = {}\n",
    "    codon_table_DNA = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',}\n",
    "    for i in range(1,7):\n",
    "        frame = i\n",
    "        codons = CODON_LIST[i]\n",
    "        protein_list = []\n",
    "        for codon in codons:\n",
    "            next_protein = codon_table_DNA[codon]\n",
    "            protein_list.append(next_protein)\n",
    "        protein_frame_dict[frame] = protein_list\n",
    "    return (protein_frame_dict)\n",
    "\n",
    "TRANSLATE (CODON_LIST)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['K', 'M', 'M', '_', 'M', '_', 'N', 'S', 'G', 'Y', 'M', 'G', 'M', 'T', 'P', 'R', 'L', 'G', 'L', 'E', 'S', 'L', 'L', 'E', '_', 'A', '_', 'M', 'I', 'R', 'V', 'A', 'S']\n",
      "-7\n",
      "-9\n",
      "-21\n",
      "['M', 'T', 'P', 'R', 'L', 'G', 'L', 'E', 'S', 'L', 'L', 'E']\n",
      "-23\n",
      "['M', 'G', 'M', 'T', 'P', 'R', 'L', 'G', 'L', 'E', 'S', 'L', 'L', 'E']\n",
      "-28\n",
      "-29\n",
      "['M']\n",
      "-30\n",
      "-31\n",
      "['M']\n",
      "-32\n",
      "['M', 'M']\n",
      "['K', '_', 'C', 'K', 'C', 'K', 'T', 'Q', 'V', 'T', 'W', 'G', '_', 'P', 'R', 'D', 'L', 'D', '_', 'S', 'L', 'F', 'W', 'N', 'K', 'P', 'E', '_', 'S', 'E', '_', 'H']\n",
      "-2\n",
      "-5\n",
      "-14\n",
      "-20\n",
      "-31\n",
      "['N', 'D', 'V', 'N', 'V', 'K', 'L', 'R', 'L', 'H', 'G', 'D', 'D', 'P', 'A', 'T', 'W', 'I', 'R', 'V', 'S', 'F', 'G', 'I', 'S', 'L', 'N', 'D', 'P', 'S', 'S', 'I']\n",
      "['L', 'R', 'C', 'Y', 'S', 'D', 'H', 'S', 'G', 'L', 'F', 'Q', 'K', 'R', 'L', '_', 'S', 'K', 'S', 'R', 'G', 'H', 'P', 'H', 'V', 'T', '_', 'V', 'L', 'H', 'L', 'H', 'H']\n",
      "-7\n",
      "-18\n",
      "['D', 'A', 'T', 'R', 'I', 'I', 'Q', 'A', 'Y', 'S', 'K', 'R', 'D', 'S', 'N', 'P', 'S', 'R', 'G', 'V', 'I', 'P', 'M', '_', 'P', 'E', 'F', 'Y', 'I', 'Y', 'I']\n",
      "-8\n",
      "-9\n",
      "['M']\n",
      "['M', 'L', 'L', 'G', 'S', 'F', 'R', 'L', 'I', 'P', 'K', 'E', 'T', 'L', 'I', 'Q', 'V', 'A', 'G', 'S', 'S', 'P', 'C', 'N', 'L', 'S', 'F', 'T', 'F', 'T', 'S']\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "next translate then search backwards through list from list[-1] if stop codon then start recording in variable \n",
    "until M is reached in which case you record the ORF and continue. if list[0] or a stop is reached then break \n",
    "and start from new stop or iterate next ORF set.\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"def ORF_FINDER(CODONS)\n",
    "    for i in range(1,7):\n",
    "        frame = i\n",
    "        codons = CODON_LIST[i]\n",
    "        protein_list = []\n",
    "        for codon in codons:\n",
    "            next_protein = codon_table_DNA[codon]\n",
    "            protein_list.append(next_protein)\n",
    "        protein_frame_dict[frame] = protein_list\n",
    "    return (protein_frame_dict)\n",
    "\"\"\"\n",
    "\n",
    "for i in range(1,7):\n",
    "    print(TRANSLATE (CODON_LIST)[i])\n",
    "    amino_acids = TRANSLATE (CODON_LIST)[i]\n",
    "    x = len (TRANSLATE (CODON_LIST)[i]) + 1\n",
    "    stop = ''\n",
    "    start = ''\n",
    "    for position in range(-1,-x,-1):\n",
    "        amino_acid = amino_acids[position]\n",
    "        #print (amino_acid) - to confirm correct aa's allocated\n",
    "        if amino_acid == '_':\n",
    "            stop = int(position)\n",
    "            print (stop)\n",
    "            continue\n",
    "        if stop == '':\n",
    "            continue\n",
    "        elif stop != '':\n",
    "            if amino_acid == 'M':\n",
    "                start = int(position)\n",
    "                print (start)\n",
    "                print (amino_acids[start:stop])\n",
    "            elif amino_acid != 'M':\n",
    "                continue\n",
    "        continue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['M', 'T', 'P', 'R', 'L', 'G', 'L', 'E', 'S', 'L', 'L', 'E']\n",
      "['M', 'G', 'M', 'T', 'P', 'R', 'L', 'G', 'L', 'E', 'S', 'L', 'L', 'E']\n",
      "['M']\n",
      "['M']\n",
      "['M', 'M']\n",
      "['M']\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "next translate then search backwards through list from list[-1] if stop codon then start recording in variable \n",
    "until M is reached in which case you record the ORF and continue. if list[0] or a stop is reached then break \n",
    "and start from new stop or iterate next ORF set.\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"def ORF_FINDER(CODONS)\n",
    "    for i in range(1,7):\n",
    "        frame = i\n",
    "        codons = CODON_LIST[i]\n",
    "        protein_list = []\n",
    "        for codon in codons:\n",
    "            next_protein = codon_table_DNA[codon]\n",
    "            protein_list.append(next_protein)\n",
    "        protein_frame_dict[frame] = protein_list\n",
    "    return (protein_frame_dict)\n",
    "\"\"\"\n",
    "\n",
    "for i in range(1,7):\n",
    "    amino_acids = TRANSLATE (CODON_LIST)[i]\n",
    "    x = len (TRANSLATE (CODON_LIST)[i]) + 1\n",
    "    stop = ''\n",
    "    start = ''\n",
    "    for position in range(-1,-x,-1):\n",
    "        amino_acid = amino_acids[position]\n",
    "        #print (amino_acid) - to confirm correct aa's allocated\n",
    "        if amino_acid == '_':\n",
    "            stop = int(position)\n",
    "            continue\n",
    "        if stop == '':\n",
    "            continue\n",
    "        elif stop != '':\n",
    "            if amino_acid == 'M':\n",
    "                start = int(position)\n",
    "                print (amino_acids[start:stop])\n",
    "            elif amino_acid != 'M':\n",
    "                continue\n",
    "        continue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['M', 'T', 'P', 'R', 'L', 'G', 'L', 'E', 'S', 'L', 'L', 'E']\n",
      "['M', 'G', 'M', 'T', 'P', 'R', 'L', 'G', 'L', 'E', 'S', 'L', 'L', 'E']\n",
      "['M']\n",
      "['M']\n",
      "['M', 'L', 'L', 'G', 'S', 'F', 'R', 'L', 'I', 'P', 'K', 'E', 'T', 'L', 'I', 'Q', 'V', 'A', 'G', 'S', 'S', 'P', 'C', 'N', 'L', 'S']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\"def ORF_FINDER(CODONS):\\n    ORFs = []\\n    for i in range(1,7):\\n        amino_acids = CODONS[i]\\n        x = len (CODONS[i]) + 1\\n        stop = ''\\n        start = ''\\n        for position in range(-1,-x,-1):\\n            amino_acid = amino_acids[position]\\n            #print (amino_acid) - to confirm correct aa's allocated\\n            if amino_acid == '_':\\n                stop = int(position)\\n                continue\\n            if stop == '':\\n                continue\\n            elif stop != '':\\n                if amino_acid == 'M':\\n                    start = int(position)\\n                    ORFs == ORFs, (amino_acids[start:stop])\\n                elif amino_acid != 'M':\\n                    continue\\n    return ORFs\\n\\nORF_FINDER (TRANSLATE (CODON_LIST))\""
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence = 'AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG'\n",
    "\n",
    "\"\"\"\n",
    "code to get a DNA sequence into a dictionary of reading frames containing codons \n",
    "these can then be iterated through attempting to create with no dependencies such as biopython.\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "reverse complement function- provides the reverse_complement when provided a DNA input as an argument\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def reverse_complement(DNA):\n",
    "    DNA = DNA.upper()\n",
    "    revcompDict = {'A':'T', 'C':'G', 'T':'A', 'G':'C'}\n",
    "    i = 1\n",
    "    x = 0\n",
    "    revseq = ''\n",
    "    revcompseq = ''\n",
    "    for i in range (1, len(DNA) + 1):\n",
    "        revseq = DNA[-i]\n",
    "        revcompseq = revcompseq + revcompDict[revseq]\n",
    "    return revcompseq\n",
    "\n",
    "reverse_complement(sequence)\n",
    "\n",
    "\"\"\"\n",
    "CODON_dictionary function- takes a DNA sequence as an input and provides a list of codons in ALL reading frames of a sequence including\n",
    "reverse complement- outputs to a dictionary of form ORF[i]=['AAA','ATG', etc] where i is the reading frame\n",
    "the reverse codons start from 4 and go to 6\n",
    "\"\"\"\n",
    "\n",
    "def CODON_LIST(DNA):\n",
    "    DNA = DNA.upper()\n",
    "    x = 0\n",
    "    y = 3\n",
    "    j = 1\n",
    "    ORFs = {}\n",
    "    codons = []\n",
    "    while y <= (len(DNA)):\n",
    "        codons.append(DNA[x:y])\n",
    "        x = x + 3\n",
    "        y = y + 3\n",
    "        if y >= len(DNA)-2:\n",
    "            ORFs[j] = codons\n",
    "            if j < 3:\n",
    "                codons = []\n",
    "                x = 0 + j\n",
    "                y = 3 + j\n",
    "                j = j + 1\n",
    "                continue\n",
    "            else:\n",
    "                break\n",
    "    \n",
    "    DNA = reverse_complement(DNA)\n",
    "    x = 0\n",
    "    y = 3\n",
    "    j = 4\n",
    "    codons = []\n",
    "    while y <= (len(DNA)):\n",
    "        codons.append(DNA[x:y])\n",
    "        x = x + 3\n",
    "        y = y + 3\n",
    "        if y >= len(DNA)-2:\n",
    "            ORFs[j] = codons\n",
    "            if j < 6:\n",
    "                codons = []\n",
    "                x = 0 + j\n",
    "                y = 3 + j\n",
    "                j = j + 1\n",
    "                continue\n",
    "            else:\n",
    "                break\n",
    "    return (ORFs)\n",
    "\n",
    "\"\"\"\n",
    "the translate function is designed to work on any LIST OF CODONS in a dictionary format where the key is an \n",
    "integer representing the reading frame from 1 - 6. therefore it has the function to check if the reading frame\n",
    "exists in the dictionary, translate the sequences to protein and output a protein sequence dictionary.\n",
    "\"\"\"\n",
    "CODON_LIST = CODON_LIST(sequence)\n",
    "\n",
    "def TRANSLATE(CODON_LIST):\n",
    "    protein_frame_dict = {}\n",
    "    codon_table_DNA = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',}\n",
    "    for i in range(1,7):\n",
    "        frame = i\n",
    "        codons = CODON_LIST[i]\n",
    "        protein_list = []\n",
    "        for codon in codons:\n",
    "            next_protein = codon_table_DNA[codon]\n",
    "            protein_list.append(next_protein)\n",
    "        protein_frame_dict[frame] = protein_list\n",
    "    return (protein_frame_dict)\n",
    "\n",
    "TRANSLATE (CODON_LIST)\n",
    "\n",
    "\"\"\"\n",
    "next translate then search backwards through list from list[-1] if stop codon then start recording in variable \n",
    "until M is reached in which case you record the ORF and continue. if list[0] or a stop is reached then break \n",
    "and start from new stop or iterate next ORF set.\n",
    "\"\"\"\n",
    "\n",
    "for i in range(1,7):\n",
    "    amino_acids = TRANSLATE (CODON_LIST)[i]\n",
    "    x = len (TRANSLATE (CODON_LIST)[i]) + 1\n",
    "    stop = ''\n",
    "    start = ''\n",
    "    for position in range(-1,-x,-1):\n",
    "        amino_acid = amino_acids[position]\n",
    "        #print (amino_acid) - to confirm correct aa's allocated\n",
    "        if amino_acid == '_':\n",
    "            stop = int(position)\n",
    "            continue\n",
    "        if stop == '':\n",
    "            continue\n",
    "        elif stop != '':\n",
    "            if amino_acid == 'M':\n",
    "                start = int(position)\n",
    "                print (amino_acids[start:stop])\n",
    "            elif amino_acid != 'M':\n",
    "                continue\n",
    "        continue\n",
    "\n",
    "\"\"\"def ORF_FINDER(CODONS):\n",
    "    ORFs = []\n",
    "    for i in range(1,7):\n",
    "        amino_acids = CODONS[i]\n",
    "        x = len (CODONS[i]) + 1\n",
    "        stop = ''\n",
    "        start = ''\n",
    "        for position in range(-1,-x,-1):\n",
    "            amino_acid = amino_acids[position]\n",
    "            #print (amino_acid) - to confirm correct aa's allocated\n",
    "            if amino_acid == '_':\n",
    "                stop = int(position)\n",
    "                continue\n",
    "            if stop == '':\n",
    "                continue\n",
    "            elif stop != '':\n",
    "                if amino_acid == 'M':\n",
    "                    start = int(position)\n",
    "                    ORFs == ORFs, (amino_acids[start:stop])\n",
    "                elif amino_acid != 'M':\n",
    "                    continue\n",
    "    return ORFs\n",
    "\n",
    "ORF_FINDER (TRANSLATE (CODON_LIST))\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['M', 'T', 'P', 'R', 'L', 'G', 'L', 'E', 'S', 'L', 'L', 'E'], ['M', 'G', 'M', 'T', 'P', 'R', 'L', 'G', 'L', 'E', 'S', 'L', 'L', 'E'], ['M'], ['M'], ['M', 'L', 'L', 'G', 'S', 'F', 'R', 'L', 'I', 'P', 'K', 'E', 'T', 'L', 'I', 'Q', 'V', 'A', 'G', 'S', 'S', 'P', 'C', 'N', 'L', 'S']]\n"
     ]
    }
   ],
   "source": [
    "sequence = 'AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG'\n",
    "\n",
    "\"\"\"\n",
    "code to get a DNA sequence into a dictionary of reading frames containing codons \n",
    "these can then be iterated through attempting to create with no dependencies such as biopython.\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "reverse complement function- provides the reverse_complement when provided a DNA input as an argument\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def reverse_complement(DNA):\n",
    "    DNA = DNA.upper()\n",
    "    revcompDict = {'A':'T', 'C':'G', 'T':'A', 'G':'C'}\n",
    "    i = 1\n",
    "    x = 0\n",
    "    revseq = ''\n",
    "    revcompseq = ''\n",
    "    for i in range (1, len(DNA) + 1):\n",
    "        revseq = DNA[-i]\n",
    "        revcompseq = revcompseq + revcompDict[revseq]\n",
    "    return revcompseq\n",
    "\n",
    "reverse_complement(sequence)\n",
    "\n",
    "\"\"\"\n",
    "CODON_dictionary function- takes a DNA sequence as an input and provides a list of codons in ALL reading frames of a sequence including\n",
    "reverse complement- outputs to a dictionary of form ORF[i]=['AAA','ATG', etc] where i is the reading frame\n",
    "the reverse codons start from 4 and go to 6\n",
    "\"\"\"\n",
    "\n",
    "def CODON_LIST(DNA):\n",
    "    DNA = DNA.upper()\n",
    "    x = 0\n",
    "    y = 3\n",
    "    j = 1\n",
    "    ORFs = {}\n",
    "    codons = []\n",
    "    while y <= (len(DNA)):\n",
    "        codons.append(DNA[x:y])\n",
    "        x = x + 3\n",
    "        y = y + 3\n",
    "        if y >= len(DNA)-2:\n",
    "            ORFs[j] = codons\n",
    "            if j < 3:\n",
    "                codons = []\n",
    "                x = 0 + j\n",
    "                y = 3 + j\n",
    "                j = j + 1\n",
    "                continue\n",
    "            else:\n",
    "                break\n",
    "    \n",
    "    DNA = reverse_complement(DNA)\n",
    "    x = 0\n",
    "    y = 3\n",
    "    j = 4\n",
    "    codons = []\n",
    "    while y <= (len(DNA)):\n",
    "        codons.append(DNA[x:y])\n",
    "        x = x + 3\n",
    "        y = y + 3\n",
    "        if y >= len(DNA)-2:\n",
    "            ORFs[j] = codons\n",
    "            if j < 6:\n",
    "                codons = []\n",
    "                x = 0 + j\n",
    "                y = 3 + j\n",
    "                j = j + 1\n",
    "                continue\n",
    "            else:\n",
    "                break\n",
    "    return (ORFs)\n",
    "\n",
    "\"\"\"\n",
    "the translate function is designed to work on any LIST OF CODONS in a dictionary format where the key is an \n",
    "integer representing the reading frame from 1 - 6. therefore it has the function to check if the reading frame\n",
    "exists in the dictionary, translate the sequences to protein and output a protein sequence dictionary.\n",
    "\"\"\"\n",
    "CODON_LIST = CODON_LIST(sequence)\n",
    "\n",
    "def TRANSLATE(CODON_LIST):\n",
    "    protein_frame_dict = {}\n",
    "    codon_table_DNA = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',}\n",
    "    for i in range(1,7):\n",
    "        frame = i\n",
    "        codons = CODON_LIST[i]\n",
    "        protein_list = []\n",
    "        for codon in codons:\n",
    "            next_protein = codon_table_DNA[codon]\n",
    "            protein_list.append(next_protein)\n",
    "        protein_frame_dict[frame] = protein_list\n",
    "    return (protein_frame_dict)\n",
    "\n",
    "TRANSLATE (CODON_LIST)\n",
    "\n",
    "\"\"\"\n",
    "next translate then search backwards through list from list[-1] if stop codon then start recording in variable \n",
    "until M is reached in which case you record the ORF and continue. if list[0] or a stop is reached then break \n",
    "and start from new stop or iterate next ORF set.\n",
    "\"\"\"\n",
    "Final_ORFs = []\n",
    "for i in range(1,7):\n",
    "    amino_acids = TRANSLATE (CODON_LIST)[i]\n",
    "    x = len (TRANSLATE (CODON_LIST)[i]) + 1\n",
    "    stop = ''\n",
    "    start = ''\n",
    "    for position in range(-1,-x,-1):\n",
    "        amino_acid = amino_acids[position]\n",
    "        #print (amino_acid) - to confirm correct aa's allocated\n",
    "        if amino_acid == '_':\n",
    "            stop = int(position)\n",
    "            continue\n",
    "        if stop == '':\n",
    "            continue\n",
    "        elif stop != '':\n",
    "            if amino_acid == 'M':\n",
    "                start = int(position)\n",
    "                Final_ORFs.append(amino_acids[start:stop])\n",
    "            elif amino_acid != 'M':\n",
    "                continue\n",
    "        continue\n",
    "print (Final_ORFs)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['MTPRLGLESLLE', 'MGMTPRLGLESLLE', 'M', 'M', 'MLLGSFRLIPKETLIQVAGSSPCNLS']\n",
      "{'M', 'MGMTPRLGLESLLE', 'MTPRLGLESLLE', 'MLLGSFRLIPKETLIQVAGSSPCNLS'}\n",
      "M MGMTPRLGLESLLE MTPRLGLESLLE MLLGSFRLIPKETLIQVAGSSPCNLS\n"
     ]
    }
   ],
   "source": [
    "sequence = 'AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG'\n",
    "\n",
    "\"\"\"\n",
    "code to get a DNA sequence into a dictionary of reading frames containing codons \n",
    "these can then be iterated through attempting to create with no dependencies such as biopython.\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "reverse complement function- provides the reverse_complement when provided a DNA input as an argument\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def reverse_complement(DNA):\n",
    "    DNA = DNA.upper()\n",
    "    revcompDict = {'A':'T', 'C':'G', 'T':'A', 'G':'C'}\n",
    "    i = 1\n",
    "    x = 0\n",
    "    revseq = ''\n",
    "    revcompseq = ''\n",
    "    for i in range (1, len(DNA) + 1):\n",
    "        revseq = DNA[-i]\n",
    "        revcompseq = revcompseq + revcompDict[revseq]\n",
    "    return revcompseq\n",
    "\n",
    "reverse_complement(sequence)\n",
    "\n",
    "\"\"\"\n",
    "CODON_dictionary function- takes a DNA sequence as an input and provides a list of codons in ALL reading frames of a sequence including\n",
    "reverse complement- outputs to a dictionary of form ORF[i]=['AAA','ATG', etc] where i is the reading frame\n",
    "the reverse codons start from 4 and go to 6\n",
    "\"\"\"\n",
    "\n",
    "def CODON_LIST(DNA):\n",
    "    DNA = DNA.upper()\n",
    "    x = 0\n",
    "    y = 3\n",
    "    j = 1\n",
    "    ORFs = {}\n",
    "    codons = []\n",
    "    while y <= (len(DNA)):\n",
    "        codons.append(DNA[x:y])\n",
    "        x = x + 3\n",
    "        y = y + 3\n",
    "        if y >= len(DNA)-2:\n",
    "            ORFs[j] = codons\n",
    "            if j < 3:\n",
    "                codons = []\n",
    "                x = 0 + j\n",
    "                y = 3 + j\n",
    "                j = j + 1\n",
    "                continue\n",
    "            else:\n",
    "                break\n",
    "    \n",
    "    DNA = reverse_complement(DNA)\n",
    "    x = 0\n",
    "    y = 3\n",
    "    j = 4\n",
    "    codons = []\n",
    "    while y <= (len(DNA)):\n",
    "        codons.append(DNA[x:y])\n",
    "        x = x + 3\n",
    "        y = y + 3\n",
    "        if y >= len(DNA)-2:\n",
    "            ORFs[j] = codons\n",
    "            if j < 6:\n",
    "                codons = []\n",
    "                x = 0 + j\n",
    "                y = 3 + j\n",
    "                j = j + 1\n",
    "                continue\n",
    "            else:\n",
    "                break\n",
    "    return (ORFs)\n",
    "\n",
    "\"\"\"\n",
    "the translate function is designed to work on any LIST OF CODONS in a dictionary format where the key is an \n",
    "integer representing the reading frame from 1 - 6. therefore it has the function to check if the reading frame\n",
    "exists in the dictionary, translate the sequences to protein and output a protein sequence dictionary.\n",
    "\"\"\"\n",
    "CODON_LIST = CODON_LIST(sequence)\n",
    "\n",
    "def TRANSLATE(CODON_LIST):\n",
    "    protein_frame_dict = {}\n",
    "    codon_table_DNA = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',}\n",
    "    for i in range(1,7):\n",
    "        frame = i\n",
    "        codons = CODON_LIST[i]\n",
    "        protein_list = []\n",
    "        for codon in codons:\n",
    "            next_protein = codon_table_DNA[codon]\n",
    "            protein_list.append(next_protein)\n",
    "        protein_frame_dict[frame] = protein_list\n",
    "    return (protein_frame_dict)\n",
    "\n",
    "AMINO_ACIDS = TRANSLATE (CODON_LIST)\n",
    "\n",
    "\"\"\"\n",
    "next translate then search backwards through list from list[-1] if stop codon then start recording in variable \n",
    "until M is reached in which case you record the ORF and continue. if list[0] or a stop is reached then break \n",
    "and start from new stop or iterate next ORF set.\n",
    "\"\"\"\n",
    "def ORF_FINDER(Amino_Acid_Dict):\n",
    "    Final_ORFs = []\n",
    "    for i in range(1,7):\n",
    "        amino_acids = Amino_Acid_Dict[i]\n",
    "        x = len (Amino_Acid_Dict[i]) + 1\n",
    "        stop = ''\n",
    "        start = ''\n",
    "        for position in range(-1,-x,-1):\n",
    "            ORF = \"\"\n",
    "            amino_acid = amino_acids[position]\n",
    "            #print (amino_acid) - to confirm correct aa's allocated\n",
    "            if amino_acid == '_':\n",
    "                stop = int(position)\n",
    "                continue\n",
    "            if stop == '':\n",
    "                continue\n",
    "            elif stop != '':\n",
    "                if amino_acid == 'M':\n",
    "                    start = int(position)\n",
    "                    ORF = (amino_acids[start:stop])\n",
    "                    joined_ORF = \"\".join(ORF)\n",
    "                    Final_ORFs.append(joined_ORF)\n",
    "                elif amino_acid != 'M':\n",
    "                    continue\n",
    "            continue\n",
    "    \n",
    "    return (Final_ORFs)\n",
    "\"\"\"\n",
    "Because of the mutable nature of the list object you can't directly use the set command (gives unordered)\n",
    "but unique list of everything in a list so first needs to be converted to a tuple then a set then back to a\n",
    "list to use the join method. unfortunate but necessary probably a more elegant way to do this.\n",
    "\"\"\"\n",
    "print (ORF_FINDER(AMINO_ACIDS))\n",
    "\n",
    "tupe = tuple (ORF_FINDER(AMINO_ACIDS))\n",
    "sets = set (tupe)\n",
    "print (sets)\n",
    "lists = list (sets)\n",
    "print (' '.join(lists))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "''",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-c31a1fb1f160>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     16\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mFASTA_DICT\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m \u001b[0mfilename\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFasta_parser\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Inputs/Example_Fasta_R99.fa'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mFasta_parser2\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mFASTA_DICT\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-13-c31a1fb1f160>\u001b[0m in \u001b[0;36mFasta_parser\u001b[0;34m(filename)\u001b[0m\n\u001b[1;32m     11\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m'>'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m             \u001b[0msequence\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m             \u001b[0mFASTA_DICT\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mseq_name\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFASTA_DICT\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mseq_name\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[0mfasta_file\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: ''"
     ]
    }
   ],
   "source": [
    "def Fasta_parser(filename):\n",
    "    fasta_file = open(filename,'r') #open file in read mode\n",
    "    FASTA_DICT = {} #initialise dictionary object\n",
    "\n",
    "    for line in fasta_file: #iterate through the lines in the fasta file\n",
    "        line = line.rstrip() #remove the newline character from the end of the line\n",
    "        seq_name = ''\n",
    "        if line[0] == '>':\n",
    "            seq_name = line[1:]\n",
    "            FASTA_DICT[seq_name] = '' # create empty entry in dictionary value \n",
    "        elif line[0] != '>':\n",
    "            sequence = line\n",
    "            FASTA_DICT[seq_name] = FASTA_DICT[seq_name] + sequence\n",
    "        \n",
    "    fasta_file.close()\n",
    "    return FASTA_DICT\n",
    "\n",
    "filename = Fasta_parser('Inputs/Example_Fasta_R99.fa')\n",
    "\n",
    "def Fasta_parser2(FASTA_DICT):\n",
    "\n",
    "    sequences_list = ''\n",
    "    SEQUENCES_LIST = []\n",
    "\n",
    "    for seq_name, sequence in FASTA_DICT.items():\n",
    "        sequences_list =  sequences_list + sequence #merges all into one long string\n",
    "        SEQUENCES_LIST.append(sequence)\n",
    "    return SEQUENCES_LIST\n",
    "                          \n",
    "sequence = Fasta_parser2(Filename)\n",
    "\n",
    "\"\"\"\n",
    "code to get a DNA sequence into a dictionary of reading frames containing codons \n",
    "these can then be iterated through attempting to create with no dependencies such as biopython.\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "reverse complement function- provides the reverse_complement when provided a DNA input as an argument\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def reverse_complement(DNA):\n",
    "    DNA = DNA.upper()\n",
    "    revcompDict = {'A':'T', 'C':'G', 'T':'A', 'G':'C'}\n",
    "    i = 1\n",
    "    x = 0\n",
    "    revseq = ''\n",
    "    revcompseq = ''\n",
    "    for i in range (1, len(DNA) + 1):\n",
    "        revseq = DNA[-i]\n",
    "        revcompseq = revcompseq + revcompDict[revseq]\n",
    "    return revcompseq\n",
    "\n",
    "reverse_complement(sequence)\n",
    "\n",
    "\"\"\"\n",
    "CODON_dictionary function- takes a DNA sequence as an input and provides a list of codons in ALL reading frames of a sequence including\n",
    "reverse complement- outputs to a dictionary of form ORF[i]=['AAA','ATG', etc] where i is the reading frame\n",
    "the reverse codons start from 4 and go to 6\n",
    "\"\"\"\n",
    "\n",
    "def CODON_LIST(DNA):\n",
    "    DNA = DNA.upper()\n",
    "    x = 0\n",
    "    y = 3\n",
    "    j = 1\n",
    "    ORFs = {}\n",
    "    codons = []\n",
    "    while y <= (len(DNA)):\n",
    "        codons.append(DNA[x:y])\n",
    "        x = x + 3\n",
    "        y = y + 3\n",
    "        if y >= len(DNA)-2:\n",
    "            ORFs[j] = codons\n",
    "            if j < 3:\n",
    "                codons = []\n",
    "                x = 0 + j\n",
    "                y = 3 + j\n",
    "                j = j + 1\n",
    "                continue\n",
    "            else:\n",
    "                break\n",
    "    \n",
    "    DNA = reverse_complement(DNA)\n",
    "    x = 0\n",
    "    y = 3\n",
    "    j = 4\n",
    "    codons = []\n",
    "    while y <= (len(DNA)):\n",
    "        codons.append(DNA[x:y])\n",
    "        x = x + 3\n",
    "        y = y + 3\n",
    "        if y >= len(DNA)-2:\n",
    "            ORFs[j] = codons\n",
    "            if j < 6:\n",
    "                codons = []\n",
    "                x = 0 + j\n",
    "                y = 3 + j\n",
    "                j = j + 1\n",
    "                continue\n",
    "            else:\n",
    "                break\n",
    "    return (ORFs)\n",
    "\n",
    "\"\"\"\n",
    "the translate function is designed to work on any LIST OF CODONS in a dictionary format where the key is an \n",
    "integer representing the reading frame from 1 - 6. therefore it has the function to check if the reading frame\n",
    "exists in the dictionary, translate the sequences to protein and output a protein sequence dictionary.\n",
    "\"\"\"\n",
    "CODON_LIST = CODON_LIST(sequence)\n",
    "\n",
    "def TRANSLATE(CODON_LIST):\n",
    "    protein_frame_dict = {}\n",
    "    codon_table_DNA = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',}\n",
    "    for i in range(1,7):\n",
    "        frame = i\n",
    "        codons = CODON_LIST[i]\n",
    "        protein_list = []\n",
    "        for codon in codons:\n",
    "            next_protein = codon_table_DNA[codon]\n",
    "            protein_list.append(next_protein)\n",
    "        protein_frame_dict[frame] = protein_list\n",
    "    return (protein_frame_dict)\n",
    "\n",
    "AMINO_ACIDS = TRANSLATE (CODON_LIST)\n",
    "\n",
    "\"\"\"\n",
    "next translate then search backwards through list from list[-1] if stop codon then start recording in variable \n",
    "until M is reached in which case you record the ORF and continue. if list[0] or a stop is reached then break \n",
    "and start from new stop or iterate next ORF set.\n",
    "\"\"\"\n",
    "def ORF_FINDER(Amino_Acid_Dict):\n",
    "    Final_ORFs = []\n",
    "    for i in range(1,7):\n",
    "        amino_acids = Amino_Acid_Dict[i]\n",
    "        x = len (Amino_Acid_Dict[i]) + 1\n",
    "        stop = ''\n",
    "        start = ''\n",
    "        for position in range(-1,-x,-1):\n",
    "            ORF = \"\"\n",
    "            amino_acid = amino_acids[position]\n",
    "            #print (amino_acid) - to confirm correct aa's allocated\n",
    "            if amino_acid == '_':\n",
    "                stop = int(position)\n",
    "                continue\n",
    "            if stop == '':\n",
    "                continue\n",
    "            elif stop != '':\n",
    "                if amino_acid == 'M':\n",
    "                    start = int(position)\n",
    "                    ORF = (amino_acids[start:stop])\n",
    "                    joined_ORF = \"\".join(ORF)\n",
    "                    Final_ORFs.append(joined_ORF)\n",
    "                elif amino_acid != 'M':\n",
    "                    continue\n",
    "            continue\n",
    "    \n",
    "    return (Final_ORFs)\n",
    "\"\"\"\n",
    "Because of the mutable nature of the list object you can't directly use the set command (gives unordered)\n",
    "but unique list of everything in a list so first needs to be converted to a tuple then a set then back to a\n",
    "list to use the join method. unfortunate but necessary probably a more elegant way to do this.\n",
    "\"\"\"\n",
    "\n",
    "tupe = tuple (ORF_FINDER(AMINO_ACIDS))\n",
    "sets = set (tupe)\n",
    "lists = list (sets)\n",
    "print (' '.join(lists))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
